sql: fix dropping zone configs from REGIONAL BY ROW during failure

otan opened the pull request:
Also added some more useful error messaging + fixed validation to use
NonDropIndexes as Active, which is in line with how the ApplyOption
function works.

Refs: #63201

Release note (bug fix): Fix a bug where REGIONAL BY ROW zone configs
were dropped before REGIONAL BY ROW changes are finalised. This causes a
bug when the REGIONAL BY ROW transformation fail.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/63200)
<!-- Reviewable:end -->

otan committed:
sql: fix dropping zone configs from REGIONAL BY ROW during failure

Also added some more useful error messaging + fixed validation to use
NonDropIndexes as Active, which is in line with how the ApplyOption
function works.

Also refactored the placeholder zone config logic when deleting the last
subzone to work in the new world.

Release note (bug fix): Fix a bug where REGIONAL BY ROW zone configs
were dropped before REGIONAL BY ROW changes are finalised. This causes a
bug when the REGIONAL BY ROW transformation fail.

ajstorm reviewed (approved):
> Reviewed 3 of 3 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/63200)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @otan)

___
*[pkg/sql/region_util.go, line 590 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/63200#-MXjQ-Hk31oODxKCR3jX:-MXjQ-HlDSuzcbiq4ob8:b-rnotkz) ([raw file](https://github.com/cockroachdb/cockroach/blob/651184b3cfbd7c6d336bb49e7152e3d48ea06416/pkg/sql/region_util.go#L590)):*
> ```Go
> 	// the partitions. If we encounter this, instead of re-writing the zone
> 	// configuration, we want to delete it.
> 	numReplicasIsZero := newZoneConfig.NumReplicas != nil && *newZoneConfig.NumReplicas == 0
> ```

Nit: Since this should now be done in the drop path, can we add an assertion here that we won't find any zone configs in this state?


<!-- Sent from Reviewable.io -->

otan reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/63200)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajstorm)

___
*[pkg/sql/region_util.go, line 590 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/63200#-MXjQ-Hk31oODxKCR3jX:-MXjR1cBDOzjhSa0Bbxv:b-pg103i) ([raw file](https://github.com/cockroachdb/cockroach/blob/651184b3cfbd7c6d336bb49e7152e3d48ea06416/pkg/sql/region_util.go#L590)):*
<details><summary><i>Previously, ajstorm (Adam Storm) wrote…</i></summary><blockquote>

Nit: Since this should now be done in the drop path, can we add an assertion here that we won't find any zone configs in this state?
</blockquote></details>

i think we can though - otherwise the

```
	if isPlaceholderZoneConfigForMultiRegion(newZoneConfig) {
		newZoneConfig.NumReplicas = proto.Int32(0)
	}

```

above won't trigger?


<!-- Sent from Reviewable.io -->

otan commented:
bors r=ajstorm

ajstorm reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/63200)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/sql/region_util.go, line 590 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/63200#-MXjQ-Hk31oODxKCR3jX:-MXlKUnK1t30fR7TUAOM:b-c8wob1) ([raw file](https://github.com/cockroachdb/cockroach/blob/651184b3cfbd7c6d336bb49e7152e3d48ea06416/pkg/sql/region_util.go#L590)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

i think we can though - otherwise the

```
	if isPlaceholderZoneConfigForMultiRegion(newZoneConfig) {
		newZoneConfig.NumReplicas = proto.Int32(0)
	}

```

above won't trigger?
</blockquote></details>

Are you saying that we can?  Or that we can't?  The "though" and "?" are confusing me.


<!-- Sent from Reviewable.io -->

otan commented:
___
*[pkg/sql/region_util.go, line 590 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/63200#-MXjQ-Hk31oODxKCR3jX:-MXlLQ654Zh497BGh7T6:bx5m3kl) ([raw file](https://github.com/cockroachdb/cockroach/blob/651184b3cfbd7c6d336bb49e7152e3d48ea06416/pkg/sql/region_util.go#L590)):*
<details><summary><i>Previously, ajstorm (Adam Storm) wrote…</i></summary><blockquote>

Are you saying that we can?  Or that we can't?  The "though" and "?" are confusing me.
</blockquote></details>

yes I meant can't. There can be zone config in this state in the eg REGIONAL BY TABLE transition 


<!-- Sent from Reviewable.io -->
